File: 224. Basic Calculator.py
class Solution:
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        ans = [[0, True]]
        positive = True
        num = ''
        i = 0
        while i < len(s):
            if s[i] == ' ':
                i += 1
                continue
            if s[i] == '-':
                positive = False
                i += 1
                continue
            if s[i] == '+':
                positive = True
                i += 1
                continue
            if s[i] == '(':
                if positive or i == 0 or s[i - 1] == '(':
                    ans.append([0, True])
                else:
                    ans.append([0, False])
                positive = True
                i += 1
                continue
            if s[i] == ')':
                if ans[-1][1]:
                    ans[-2][0] += ans[-1][0]
                    ans.pop()
                else:
                    ans[-2][0] -= ans[-1][0]
                    ans.pop()
                i += 1
                continue
            while i < len(s) and ord('0') <= ord(s[i]) <= ord('9'):
                num += s[i]
                i += 1
                if i == len(s) or (i < len(s) and not ord('0') <= ord(s[i]) <= ord('9')):
                    num = int(num)
                    if positive:
                        ans[-1][0] += num
                    else:
                        ans[-1][0] -= num
                    i -= 1
                    break
            else:
                i -= 1
            i += 1
            num = ''
        return ans[0][0]


sol = Solution()
print(sol.calculate("1+( 5 - ( 5 ) + 2) - 2"))

File: 224. Basic Calculator.cpp
#include <iostream>
#include <stack>
using namespace std;

class Solution {
public:
    int calculate(string s) {
        stack<long long> st;
        long long ans = 0, num = 0;
        int sign = +1;  // positive +1 negative -1
        for (char ch: s) {
            switch (ch) {
            case ' ':  // ignore space
                break;
            case '+':
                ans += num * sign;
                sign = +1;
                num = 0;
                break;
            case '-':
                ans += num * sign;
                sign = -1;
                num = 0;
                break;
            case '(':
                st.push(ans);
                st.push((long long)sign);
                ans = 0;
                num = 0;
                sign = +1;
                break;
            case ')':
                ans += num * sign;
                sign = st.top();
                st.pop();
                ans = st.top() + sign * ans;
                st.pop();
                sign = +1;
                num = 0;
                break;
            default:  // digit
                num = num * 10 + (ch - '0');
                break;
            }
        }
        return ans + sign * num;
    }
};

int main() {
    Solution a;
    cout << a.calculate("1+2+3+4+5") << endl;
    cout << a.calculate("1+2+3+(4+5)") << endl;
    cout << a.calculate("1+(-2)+3+(4+5)") << endl;
    cout << a.calculate("1-(2+2+4)+5") << endl;
    cout << a.calculate("1+2+3-(4+5)") << endl;
    cout << a.calculate("(-1+2+3-4+5)") << endl;
    return 0;
}
